src/__internal.rs

```diff
@@ -51,7 +51,10 @@ where
 ///
 /// # Safety
 ///
-/// Only the `init` module is allowed to use this trait.
+/// Implementers (typically the `#[pin_data]` macro) must ensure that this method
+/// returns a `PinData` instance that accurately reflects the structural pinning
+/// and field layout of `Self`. Incorrect metadata can lead to undefined behavior
+/// when used by the `pin-init` macros.
```

Comment:
This trait should only be implement by the `#[pin_data]` macro. Same for the other traits in this file. Also please use bullet points for safety sections with multiple requirements (one requirement per point). I'm not sure we want all of the internal workings written down in the safety comment here. I don't really see the value in that.